targets/TARGET_STM/us_ticker_16b.c

```diff
@@ -91,22 +90,25 @@ uint32_t us_ticker_read()
 
 void us_ticker_set_interrupt(timestamp_t timestamp)
 {
-    int delta = (int)((uint32_t)timestamp - us_ticker_read());
-
-    uint16_t cval = TIM_MST->CNT;
+    int current_time = us_ticker_read();
+    int delta = (int)(timestamp - current_time);
 
     if (delta <= 0) { // This event was in the past
-        us_ticker_irq_handler();
+        /* Force the event to be handled in next interrupt context
+         * This prevents calling interrupt handlers in loops as
+         * us_ticker_set_interrupt might called again from the
+         * application handler
+         */
+        oc_int_part = 0;
+        TimMasterHandle.Instance = TIM_MST;
+        HAL_TIM_GenerateEvent(&TimMasterHandle, TIM_EVENTSOURCE_CC1);
     } else {
+        /*  set the comparator at the timestamp lower 16 bits
+         *  and count the number of wrap-around loops to do with
+         *  the upper 16 bits
+         */
         oc_int_part = (uint32_t)(delta >> 16);
-        oc_rem_part = (uint16_t)(delta & 0xFFFF);
-        if (oc_rem_part <= (0xFFFF - cval)) {
-            set_compare(cval + oc_rem_part);
-            oc_rem_part = 0;
-        } else {
-            set_compare(0xFFFF);
-            oc_rem_part = oc_rem_part - (0xFFFF - cval);
-        }
+        set_compare(timestamp & 0xFFFF);
     }
```

Comment:
Ok that took me longer to answer here because I need to read everything and I actually have run tests as well, with varying timeouts from 0 to 0X10000 and more and have tried to detect possible hick-ups.   I tested your above code and the alternative of mine with the 32bits sequencing, both are running fine with the CPU @ 48MHz.  

I'll go with your code which is well thought and very described as well.
I've been through the various cases you've listed and timings I could thought of, and could not find a failure case so far. + the tests I've run are all fine.

I'll restart the basic mbed test suite as well and post the results 

Thanks a lot for this effort !